modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/consensus/ValidationErrorStorage.scala

```diff
@@ -0,0 +1,54 @@
+package io.constellationnetwork.node.shared.infrastructure.consensus
+
+import java.time.Instant
+
+import cats.data.NonEmptyList
+import cats.effect.kernel.{Async, Clock, Ref}
+import cats.effect.std.Semaphore
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+trait ValidationErrorStorage[F[_], Artifact, ValidationError] {
+  def addValidationError(artifact: Artifact, errors: ValidationError): F[Unit]
+  def addValidationErrors(errors: NonEmptyList[(Artifact, ValidationError)]): F[Unit]
+  def allValidationErrors: F[Vector[(Artifact, ValidationError, Instant)]]
+}
+
+object ValidationErrorStorage {
+
+  def make[F[_]: Async, Artifact, ValidationError](
+    maxSize: Int
+  )(implicit clock: Clock[F]): F[ValidationErrorStorage[F, Artifact, ValidationError]] =
+    for {
+      updateSemaphore <- Semaphore[F](1)
+      validationErrorsR <- Ref[F].of(Vector.empty[(Artifact, ValidationError, Instant)])
+    } yield
+      new ValidationErrorStorage[F, Artifact, ValidationError] {
+        override def addValidationError(artifact: Artifact, errors: ValidationError): F[Unit] =
+          updateSemaphore.permit.use { _ =>
```

Comment:
Do we need to use semaphore? As far as I know, the `Ref` itself already provides safe concurrent access and modification (non-blocking): https://typelevel.org/cats-effect/docs/std/ref